File: buque/core/writer.py
from __future__ import annotations

from pathlib import Path

import fitz

from buque.core.models import TocNode


def write_bookmarks(input_path: Path, output_path: Path, toc_nodes: list[TocNode]) -> None:
    output_path.parent.mkdir(parents=True, exist_ok=True)
    same_file = input_path.resolve() == output_path.resolve()
    with fitz.open(input_path) as doc:
        toc_payload = [[node.level, node.title, node.page_index + 1] for node in toc_nodes]
        doc.set_toc(toc_payload)
        if same_file:
            temp_path = output_path.with_suffix(f"{output_path.suffix}.tmp")
            doc.save(temp_path)
            temp_path.replace(output_path)
        else:
            doc.save(output_path)

File: src/core/writer.ts
import { copyFile, mkdir, readFile, rename, rm, writeFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { outlinePdfFactory } from "@lillallol/outline-pdf";
import * as pdfLib from "pdf-lib";
import type { TocNode } from "../types.js";

export async function writeBookmarks(options: { inputPath: string; outputPath: string; tocNodes: TocNode[] }): Promise<void> {
  await mkdir(dirname(options.outputPath), { recursive: true });
  const sameFile = resolve(options.inputPath) === resolve(options.outputPath);
  const targetPath = sameFile ? `${options.outputPath}.tmp` : options.outputPath;
  try {
    if (options.tocNodes.length === 0) {
      await copyFile(options.inputPath, targetPath);
    } else {
      const pdf = await readFile(options.inputPath);
      const outline = outlineText(options.tocNodes);
      const outlinePdf = outlinePdfFactory(pdfLib);
      const outlinedDocument = await outlinePdf({ pdf, outline });
      await writeFile(targetPath, await outlinedDocument.save());
    }
    if (sameFile) await rename(targetPath, options.outputPath);
  } catch (error) {
    if (sameFile) await rm(targetPath, { force: true });
    throw error;
  }
}

function outlineText(tocNodes: TocNode[]): string {
  return tocNodes
    .map((node) => {
      const pageNumber = Math.max(1, node.pageIndex + 1);
      const depth = "-".repeat(Math.max(0, node.level - 1));
      return `${pageNumber}|${depth}|${node.title}`;
    })
    .join("\n");
}
